src/application_settings/container_section_base.py

```diff
@@ -79,7 +77,16 @@ def _set(self) -> Self:
 
     def _update(self, changes: dict[str, Any]) -> Self:
         "Update parameters and sections with data specified in changes; not meant for config"
-        return replace(self, **changes)
+        # in self._set(), which normally is always executed, we ensured that
+        # self is a dataclass instance
+        return replace(self, **changes)  # type: ignore[type-var]
+
+
+def _check_dataclass_decorator(obj: Any) -> None:
+    if not (is_dataclass(obj)):
+        raise TypeError(
+            f"{obj} is not a dataclass; did you forget to add '@dataclass(frozen=True)' when you defined {obj.__class__}?."
```

Comment:
Can and should we check the frozen tag as well?

Reply:
Right, good idea.
Done.